src/v/io/io_queue.cc

```diff
@@ -0,0 +1,197 @@
+
+/*
+ * closing the file always succeeds. once seastar::file::close is invoked the
+ * handle is not longer usable, so we swallow any errors here and release units
+ * to avoid blocking other queues as there isn't much else we can do.
+ *
+ * failure to close a file is quite rare outside of the normal cases like trying
+ * to close an unknown or already closed file handle. these cases won't occur
+ * using the seastar api since it prevents those mistakes. in any case, what we
+ * care about most when closing a file is that its file descriptor resource is
+ * freed in the kernel to avoid reaching an open file limit.
+ */
+seastar::future<> io_queue::close() noexcept {
+    log.debug("Closing {}", path_);
+    assert(opened());
+
+    // ensure all on-going ops have completed
+    auto units = co_await seastar::get_units(
+      ops_, seastar::semaphore::max_counter());
+
+    auto file = std::exchange(file_, nullptr);
+    try {
+        co_await file->close();
+    } catch (...) {
+        log.warn(
+          "Unable to close file {}: {}", path_, std::current_exception());
+    }
+}
+
+void io_queue::submit_read(page& page) noexcept {
+    page.set_flag(page::flags::read);
+    enqueue_pending(page);
+}
+
+void io_queue::submit_write(page& page) noexcept {
+    /*
+     * if the page is not on any list, then in queue it.
+     */
+    if (!page.io_queue_hook.is_linked()) {
+        page.set_flag(page::flags::write);
+        enqueue_pending(page);
+        return;
+    }
+
+    assert(page.test_flag(page::flags::write));
+
+    /*
+     * if it isn't queued, then that means the dispatcher has handled it. in
+     * this case, new data may have arrived into the page. we mark it as queued
+     * as a signal to the I/O completion continuation to requeue the write.
+     */
+    if (!page.test_flag(page::flags::queued)) {
+        log.debug("Marking inflight I/O to be requeued");
+        page.set_flag(page::flags::queued);
+    }
+}
+
+void io_queue::enqueue_pending(page& page) noexcept {
+    pending_.push_back(page);
+    page.set_flag(page::flags::queued);
+    cond_.signal();
+}
+
+seastar::future<> io_queue::dispatch() noexcept {
+    while (true) {
+        co_await cond_.wait([this] { return !pending_.empty() || stop_; });
+
+        if (stop_) {
+            co_return;
+        }
+
+        auto units = co_await seastar::get_units(ops_, 1);
+
+        if (file_ == nullptr) {
+            continue;
+        }
+
+        auto& page = pending_.front();
+        pending_.pop_front();
+        assert(page.test_flag(page::flags::queued));
+        page.clear_flag(page::flags::queued);
+        running_.push_back(page);
+
+        if (page.test_flag(page::flags::read)) {
+            dispatch_read(page, std::move(units));
+
+        } else if (page.test_flag(page::flags::write)) {
+            dispatch_write(page, std::move(units));
+
+        } else {
+            assert(false);
+            std::terminate();
```

Comment:
nit: maybe this is a placeholder/never supposed to execute but should there be a vassert or vlog here to add some context on why this will fail? The `std::terminate();` statement is unreachable.

Reply:
> The std::terminate(); statement is unreachable.

Well in debug mode `assert` is on right? Anyways +1 to vassert here

Reply:
I added vassert for now. This is super hot path and vassert is slow, but I can let perf results guide us :)

Reply:
I mean if it's the super hot path we should not be using coroutines right? :)

+1 to profiling results guiding us.

Reply:
Oh, the dispatch loop can be coroutine and have a hot path inside of it. I thought i had a tight loop over the request queue intrusive list, but apparently not. As written currently you're right it's slower than necessary.

Reply:
maybe just add an extra if (...) to avoid vassert performance penalty (and we should probably fix vassert if it has measurable impact).

```
if (!cond) {
  vassert(cond, "err");
}
```

Reply:
I read this and assumed that the check is slow for some reason
> This is super hot path and vassert is slow